binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+The data remain on-disk, and are not loaded into memory.
+Nevertheless, we are able to call dplyr functions to interact with the data:
+
+```{r glimpse-data}
+#| cache: true
+glimpse(seattle_lib)
+```
+
+The first line in the output tell you that `seattle_lib` is stored locally on-disk as a single csv file.
+The second line indicates that the data are stored as a tabular data set with 41 million rows and 12 columns.
+
+### Using dplyr with arrow
+
+Writing dplyr code for arrow data is conceptually similar to dbplyr, described in @sec-import-databases.
+The dplyr code that you write in R is internally transformed to a query that the Apache Arrow C++ library understands, which is then executed when you call `collect()`.
+Here's an example that counts the total number of books checked out in each year:
+
+```{r books-by-year}
+#| cache: true
+seattle_lib |> 
+  filter(MaterialType == "BOOK") |>
+  group_by(CheckoutYear) |>
+  summarise(TotalCheckouts = sum(Checkouts)) |>
+  arrange(desc(CheckoutYear)) |>
+  collect()
+```
+
+Only some R expressions are understood by arrow, so you will not necessarily be able to write exactly the same code you usually would.
+However, the list of operations and functions supported is fairly extensive and growing.
+A complete list is provided in the arrow package documentation at [arrow.apache.org/docs/r/reference/acero.html](https://arrow.apache.org/docs/r/reference/acero.html).
+
+## The parquet file format
+
+When working with small data sets, csv files can be convenient: they're human readable plain text files, and most analysis tools can read data from a csv.
+However, there are some drawbacks:
+
+-   Ambiguity.
+    A csv file does not uniquely specify the data structure.
+    For example, a csv reader has to guess whether `"2022-10-08"` should be parsed as a string or a date.
+    If it is a date, it has to guess whether the date refers to October 8th or August 10th.
+    Here is a slightly contrived example.
+    Here is a small tibble that contains some integers and some strings that look suspiciously like dates:
+
+    ```{r csv-ambiguity-example}
+    original <- tibble(
+      integers = 1L:3L,
+      strings = c("2022-01-01", "1980-12-01", "1900-12-12")
+    )
+    original
+    ```
+
+    Simple though it is, this data set poses problems for a csv file.
+    The csv version of this data does not explicitly specify that the first column contains integers and the second column contains strings.
+    This produces ambiguity because the csv reader guesses that the data types are double and character respectively:
+
+    ```{r csv-ambiguity-failure}
+    csv_file <- tempfile(fileext = ".csv")
+    write_csv(original, csv_file)
+    read_csv(csv_file, show_col_types = FALSE)
+    ```
+
+    The `read_csv()` and `write_csv()` functions are designed to avoid this issue to an extent, but it is impossible to avoid it completely because the csv format is inherently ambiguous.
+
+-   Inaccuracy.
```

Comment:
I'd make this the last bullet, since it'd be unusual for this to make any practical difference.